lib/Basics/HashSet.h

```diff
@@ -145,19 +145,87 @@ class HashSet {
     insert(other.cbegin(), other.cend());
   }
 
-  HashSet(HashSet&& other) {
-    *this = std::move(other);
+  HashSet(HashSet&& other) noexcept {
+    if (other._buffer == nullptr) {
+      _buffer = nullptr;
+      _states = nullptr;
+      _keys = nullptr;
+    } else if (other._buffer != &other._local_buffer[0]) {
+      // we can just steal the other's buffer
+      _buffer = other._buffer;
+      _states = other._states;
+      _keys = other._keys;
+    } else {
+      // we can copy the other's local buffer
+      // TODO: make this work with non-POD types
+      memcpy(&_local_buffer[0], &other._local_buffer[0], sizeof(_local_buffer));
+      _buffer = &_local_buffer[0];
+      _states = (State*)_buffer;
+      size_t states_size = ((other._num_buckets * sizeof(State) + 8 - 1) / 8) * 8;
+      _keys  = (KeyT*)(_buffer + states_size);
+    }
+
+    _num_buckets = other._num_buckets;
+    _num_filled = other._num_filled;
+    _max_probe_length = other._max_probe_length;
+    _mask = other._mask;
+    
+    other._buffer = nullptr;
+    other._states = nullptr;
+    other._keys = nullptr;
+    other._num_buckets = 0;
+    other._num_filled = 0;
+    other._max_probe_length = -1;
+    other._mask = 0;
   }
 
   HashSet& operator=(const HashSet& other) {
-    clear();
-    reserve(other.size());
-    insert(other.cbegin(), other.cend());
+    if (this != &other) {
+      clear();
+      reserve(other.size());
+      insert(other.cbegin(), other.cend());
+    }
     return *this;
   }
 
-  void operator=(HashSet&& other) {
-    this->swap(other);
+  HashSet& operator=(HashSet&& other) noexcept {
+    if (this != &other) {
+      if (_buffer != &_local_buffer[0]) {
+        delete[] _buffer;
+      }
+      if (other._buffer == nullptr) {
+        _buffer = nullptr;
+        _states = nullptr;
+        _keys = nullptr;
+      } else if (other._buffer != &other._local_buffer[0]) {
+        // we can just steal the other's buffer
+        _buffer = other._buffer;
+        _states = other._states;
+        _keys = other._keys;
+      } else {
+        // we can copy the other's local buffer
+        // TODO: make this work with non-POD types
+        memcpy(&_local_buffer[0], &other._local_buffer[0], sizeof(_local_buffer));
```

Comment:
since currently HashSet can't handle non-POD types correctly, let's add corresponding check for that:
```
static_assert(std::is_pod<value_type>::value, "value_type must be a POD type");
```